src/content/en/updates/2018/06/audio-worklet-design-pattern.md

```diff
@@ -0,0 +1,414 @@
+class in the example code handles this operation nicely.
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-3.png">
+  <figcaption>
+    HeapAudioBuffer class for the easier usage of WASM heap
+  </figcaption>
+</figure>
+
+There is an [early proposal](https://github.com/WebAudio/web-audio-api/issues/1384)
+under discussion to integrate the WASM heap directly into the AudioWorklet
+system. Getting rid of this redundant data cloning between the JS memory and the
+WASM heap seems natural, but the specific details need to be worked out.
+
+### Handling Buffer Size Mismatch
+
+  - [Example code](https://googlechromelabs.github.io/web-audio-samples/audio-worklet/design-pattern/awn-ring-buffer/)
+
+An AudioWorkletNode and AudioWorkletProcessor pair is designed to work like a
+regular AudioNode; AudioWorkletNode handles the interaction with other codes
+while AudioWorkletProcessor takes care of internal audio processing. Because a
+regular AudioNode processes 128
+[frames](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API/Basic_concepts_behind_Web_Audio_API#Audio_buffers_frames_samples_and_channels)
+at a time, AudioWorkletProcessor must do the same to become a first-class
+citizen. This is one of the advantages of the AudioWorklet design that ensures
+no additional latency due to internal buffering is introduced within the
+AudioWorkletProcessor, but it can be a problem if a processing function requires
+a buffer size different than 128 frames. The common solution for such case is to
+use a ring buffer, also known as a circular buffer or a FIFO.
+
+Here's a diagram of AudioWorkletProcessor using two ring buffers inside to
+accommodate a WASM function that takes 512 frames in and out. (The number 512
+here is arbitrarily picked.)
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-4.png">
+  <figcaption>
+    Using RingBuffer inside of AudioWorkletProcessor's `process()` method
+  </figcaption>
+</figure>
+
+The algorithm for diagram would be:
+
+  1. AudioWorkletProcessor pushes 128 frames into the **Input RingBuffer** from
+    its Input.
+  1. Perform the following steps only if the **Input RingBuffer** has greater
+    than or equal to 512 frames.
+     1. Pull 512 frames from the **Input RingBuffer**.
+     1. Process 512 frames with the given WASM function.
+     1. Push 512 frames to the **Output RingBuffer**.
+  1. AudioWorkletProcessor pulls 128 frames from the **Output RingBuffer** to
+    fill its **Output**.
+
+As shown in the diagram, Input frames always get accumulated into Input
+RingBuffer and it handles buffer overflow by overwriting the oldest frame block
+in the buffer. That is a reasonable thing to do for a real-time audio
+application. Similarly, the Output frame block will always get pulled by the
+system. Buffer underflow (not enough data) in Output RingBuffer will result
```

Comment:
"Output Ringbuffer" -> "**Output RingBuffer**"?

"will result" -> "will result in"